pkg/cgroups/cgroupstats.go

```diff
@@ -0,0 +1,542 @@
+
+	// Files look like this:
+	//
+	// 124
+
+	usageFiles, err := filepath.Glob(path.Join(cgroupPath, "/hugetlb.*.usage_in_bytes"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	maxUsageFiles, err := filepath.Glob(path.Join(cgroupPath, "/hugetlb.*.max_usage_in_bytes"))
+
+	if err != nil {
+		return nil, err
+	}
+
+	if len(usageFiles) != len(maxUsageFiles) {
+		return nil, fmt.Errorf("error finding files")
+	}
+
+	result := make([]HugetlbUsage, len(usageFiles))
+
+	for i, file := range usageFiles {
+		tokens := hugetlbUsageInBytesRe.FindStringSubmatch(file)
+		if tokens == nil || len(tokens) != 2 {
+			return nil, fmt.Errorf("error finding files")
+		}
+		size := string(tokens[1])
+
+		result[i].Size = size
+
+		number, err := readCgroupSingleNumber(file)
+
+		if err != nil {
+			return nil, err
+		}
+
+		result[i].Bytes = number
+	}
+
+	for _, file := range maxUsageFiles {
+		tokens := hugetlbMaxUsageInBytesRe.FindStringSubmatch(file)
+		if tokens == nil || len(tokens) != 2 {
+			return nil, fmt.Errorf("error finding files")
+		}
+		size := string(tokens[1])
+
+		// Find the already existing result.
+
+		var res *HugetlbUsage
+
+		for j, tmp := range result {
+			if tmp.Size == size {
+				res = &result[j]
+			}
+		}
+		if res == nil {
+			return nil, fmt.Errorf("error finding files")
+		}
+
+		number, err := readCgroupSingleNumber(file)
+
+		if err != nil {
+			return nil, err
+		}
+
+		res.MaxBytes = number
+	}
+
+	return result, nil
+}
+
+// GetMemoryUsage retrieves cgroup memory usage.
+func GetMemoryUsage(cgroupPath string) (MemoryUsage, error) {
+
+	// Files look like this:
+	//
+	// 142
+
+	usage, err := readCgroupSingleNumber(path.Join(cgroupPath, "memory.usage_in_bytes"))
+	if err != nil {
+		return MemoryUsage{}, err
+	}
+
+	maxUsage, err := readCgroupSingleNumber(path.Join(cgroupPath, "memory.max_usage_in_bytes"))
+	if err != nil {
+		return MemoryUsage{}, err
+	}
+
+	result := MemoryUsage{
+		Bytes:    usage,
+		MaxBytes: maxUsage,
+	}
+
+	return result, nil
+}
+
+var numaLineRe = regexp.MustCompile(`^(\S+)=(\d+)([ N\d+=\d+]+)$`)
+var numaNodesRe = regexp.MustCompile(`(N\d+)=(\d+)`)
+
+// GetNumaStats returns parsed cgroup NUMA statistics.
+func GetNumaStats(cgroupPath string) (NumaStat, error) {
+
+	// File looks like this:
+	//
+	// total=44611 N0=32631 N1=7501 N2=1982 N3=2497
+	// file=44428 N0=32614 N1=7335 N2=1982 N3=2497
+	// anon=183 N0=17 N1=166 N2=0 N3=0
+	// unevictable=0 N0=0 N1=0 N2=0 N3=0
+	// hierarchical_total=768133 N0=509113 N1=138887 N2=20464 N3=99669
+	// hierarchical_file=722017 N0=496516 N1=119997 N2=20181 N3=85323
+	// hierarchical_anon=46096 N0=12597 N1=18890 N2=283 N3=14326
+	// hierarchical_unevictable=20 N0=0 N1=0 N2=0 N3=20
+
+	lines, err := readCgroupFileLines(path.Join(cgroupPath, "memory.numa_stat"))
+
+	if err != nil {
+		return NumaStat{}, err
+	}
+
+	result := NumaStat{}
+	for _, line := range lines {
+		tokens := numaLineRe.FindStringSubmatch(line)
```

Comment:
I'd prolly not use regexps here, but, just go just with `strings.Split()` and a helper function for parsing `key=value` type tokens. But I guess this is fine if it works here :smile:

Reply:
Removed regexp usage.